scripts/collect.py

```diff
@@ -0,0 +1,136 @@
+#!/usr/bin/env python3
+from math import log, tan, pi
+from itertools import product
+from argparse import ArgumentParser
+from os.path import join
+
+import tempfile, shutil, urllib.request, io, sys, subprocess
+import unittest, unittest.mock as mock
+
+tile_url = 'https://terrain-preview.mapzen.com/geotiff/{z}/{x}/{y}.tif'
+
+def mercator(lat, lon, zoom):
+    ''' Convert latitude, longitude to z/x/y tile coordinate at given zoom.
+    '''
+    # convert to radians
+    x1, y1 = lon * pi/180, lat * pi/180
+    
+    # project to mercator
+    x2, y2 = x1, log(tan(0.25 * pi + 0.5 * y1))
+    
+    # transform to tile space
+    tiles, diameter = 2 ** zoom, 2 * pi
+    x3, y3 = int(tiles * (x2 + pi) / diameter), int(tiles * (pi - y2) / diameter)
+    
+    return zoom, x3, y3
+
+def tiles(zoom, lat1, lon1, lat2, lon2):
+    ''' Convert geographic bounds into a list of tile coordinates at given zoom.
+    '''
+    # convert to geographic bounding box
+    minlat, minlon = min(lat1, lat2), min(lon1, lon2)
+    maxlat, maxlon = max(lat1, lat2), max(lon1, lon2)
+    
+    # convert to tile-space bounding box
+    _, xmin, ymin = mercator(maxlat, minlon, zoom)
+    _, xmax, ymax = mercator(minlat, maxlon, zoom)
+    
+    # generate a list of tiles
+    xs, ys = range(xmin, xmax+1), range(ymin, ymax+1)
+    tiles = [(zoom, x, y) for (y, x) in product(ys, xs)]
+    
+    return tiles
+
+def download(output_tif, tiles, verbose=True):
+    ''' Download list of tiles to a temporary directory and return its name.
+    '''
+    dir = tempfile.mkdtemp(prefix='collect-')
```

Comment:
It is associated with documentation and “consumption” tasks (not server
“assembly" tasks). Documentation is currently proposed in a new docs/ dir.
Sounds like it makes more sense to put this script into a new docs/scripts/
dir.

On Tue, Aug 23, 2016 at 9:43 AM, migurski <notifications@github.com> wrote:

> In scripts/collect.py
> <https://github.com/tilezen/joerd/pull/90#discussion_r75904932>:
>
> > +    maxlat, maxlon = max(lat1, lat2), max(lon1, lon2)
> > +
> > +    # convert to tile-space bounding box
> > +    _, xmin, ymin = mercator(maxlat, minlon, zoom)
> > +    _, xmax, ymax = mercator(minlat, maxlon, zoom)
> > +
> > +    # generate a list of tiles
> > +    xs, ys = range(xmin, xmax+1), range(ymin, ymax+1)
> > +    tiles = [(zoom, x, y) for (y, x) in product(ys, xs)]
> > +
> > +    return tiles
> > +
> > +def download(output_tif, tiles, verbose=True):
> > +    ''' Download list of tiles to a temporary directory and return its name.
> > +    '''
> > +    dir = tempfile.mkdtemp(prefix='collect-')
>
> That’s a good question! Nathaniel suggested I put it in this repo, but it
> sounds like it’s intended for standalone use from the documentation on
> mapzen.com.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tilezen/joerd/pull/90/files/37a7838e127784b16a958dfd8e58d407ebef8d4d#r75904932>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA0EO2yYMSU80UAy4dWO3wmO_434goc0ks5qiyMggaJpZM4JqXwN>
> .
>